File: src/lib/wow/formats/adt/loader-generics.ts
/**
 * Loader generics, ported from wow.export (src/js/3D/loaders/LoaderGenerics.js).
 */
import { BufferWrapper } from '../buffer';

/** Process a null-terminated string block. */
export function ReadStringBlock(data: BufferWrapper, chunkSize: number): Record<number, string> {
  const chunk = data.readBuffer(chunkSize, false) as Buffer;
  const entries: Record<number, string> = {};

  let readOfs = 0;
  for (let i = 0; i < chunkSize; i++) {
    if (chunk[i] === 0x0) {
      // Skip padding bytes.
      if (readOfs === i) {
        readOfs += 1;
        continue;
      }

      entries[readOfs] = chunk.toString('utf8', readOfs, i).replace(/\0/g, '');
      readOfs = i + 1;
    }
  }

  return entries;
}

export default { ReadStringBlock };

File: golang/internal/wow/formats/adt/loader_generics.go
package adt

import (
	"strings"

	"github.com/pqhuy98/wow-converter/internal/buffer"
)

// ReadStringBlock parses a null-terminated string block.
func ReadStringBlock(data *buffer.Buffer, chunkSize int) map[int]string {
	raw := data.ReadBuffer(buffer.ReadBufferOptions{Length: chunkSize, Wrap: false}).([]byte)
	entries := make(map[int]string)
	readOfs := 0
	for i := 0; i < chunkSize; i++ {
		if raw[i] == 0 {
			if readOfs == i {
				readOfs++
				continue
			}
			entries[readOfs] = strings.ReplaceAll(string(raw[readOfs:i]), "\x00", "")
			readOfs = i + 1
		}
	}
	return entries
}
